### PR TITLE
Test unicode symbols

### LINKED ISSUE #1191: Test unicode symbols
Add test for opening a realm with unicode symbols in the path.

Add test for queries and subscriptions with unicode parameters.

PR - [https://github.com/realm/realm-dart/pull/1186](https://github.com/realm/realm-dart/pull/1186)

### PR DESCRIPTION
**Pull request by desistefanova**
Add new tests with unicode symbols.
Generated new strings for test emails using the allowed symbols for email.
Fixes #1191
Depends on core release:
https://github.com/realm/realm-core/issues/6350
https://github.com/realm/realm-core/issues/6336

### CODE REVIEW

File: test/test.dart [hunk 1]
```diff
@@ -399,9 +399,18 @@ String generateRandomRealmPath() {
 }
 
 final random = Random();
+String _getRandom(int len, String chars) => List.generate(len, (index) => chars[random.nextInt(chars.length)]).join();
+
 String generateRandomString(int len) {
-  const chars = 'abcdefghjklmnopqrstuvwxuz';
-  return List.generate(len, (index) => chars[random.nextInt(chars.length)]).join();
+  return _getRandom(len, 'abcdefghjklmnopqrstuvwxuz');
```
Comment (blagoev) [inline test/test.dart:405]: Not sure I understand the reasoning behind this change.

File: test/test.dart [hunk 2]
```diff
@@ -399,9 +399,18 @@ String generateRandomRealmPath() {
 }
 
 final random = Random();
+String _getRandom(int len, String chars) => List.generate(len, (index) => chars[random.nextInt(chars.length)]).join();
```
Comment (blagoev) [inline test/test.dart:402]: _getRandom as a method name that returns a `String` is not great imo. I suggest we revert this change

File: test/test.dart [hunk 3]
```diff
@@ -399,9 +399,18 @@ String generateRandomRealmPath() {
...
-  return List.generate(len, (index) => chars[random.nextInt(chars.length)]).join();
+  return _getRandom(len, 'abcdefghjklmnopqrstuvwxuz');
+}
+
+String generateRandomUnicodeString(int len) {
+  return _getRandom(len, 'uvwxuzфоо-барΛορεμლორემ植物החללجمعتsøren');
+}
+
+String generateRandomEmail(int len) {
+  return _getRandom(len, r"abcdefghjklmnopqrstuvwxuz!#$%&*+-'/=?^_`{|}~0123456789");
```
Comment (blagoev) [inline test/test.dart:413]: This does not seem to generate an random email. It just generates a random String from what it seems allowed characters presumably for valid email address. 
I suggest we refactor this code to really return a random email address in full including the domain name and make this one invoke the generateRandomString

File: test/realm_test.dart [hunk 4]
```diff
@@ -1859,6 +1859,46 @@ Future<void> main([List<String>? args]) async {
     expect(realmCore.getDeviceName(), matcher);
     expect(realmCore.getDeviceVersion(), matcher);
   });
+
+  test('Realm path with unicode symbols', () {
+    var config = Configuration.local([Car.schema], path: "${generateRandomUnicodeString(10)}.realm");
```
Comment (blagoev) [inline test/realm_test.dart:1864]: This should just call generateRandomEmail passing an optional argument for including unicode symbols or not

**Review by blagoev — COMMENTED** [review]
the tests seem fine but the `generate` method changes needs refactoring

File: test/test.dart (hunk 3 above)
Comment (blagoev) [reply test/test.dart:413]: also I don't think we need all the allowed symbols in an email. We just need to make sure we are using a different email in our tests.

Comment (desistefanova) [reply test/test.dart:413]: I refactored the methods as you suggested. I also added an options for including the allowed characters for email, because I think that it is good to have them in case some email validation issue appear in future.

File: test/test.dart (hunk 2 above)
Comment (desistefanova) [reply test/test.dart]: I removed this method.

File: test/test.dart (hunk 1 above)
Comment (desistefanova) [reply test/test.dart]: I needed to generate strings containing unicode and not containing unicode symbols. That's why this method was refactored.

**Comment by desistefanova** [comment]
> One thing. There is not guarantee that the `generateRandomString` function actually return a string with non-ascii characters, when the `unicode` parameter is set to `true`. It is likely, but not guaranteed. Also, ascii characters are part of unicode, so perhaps the flag could have a better name.

Ok, I changed the name to `allowUnicode`, which means not only ASCI characters could be expected.

File: test/test.dart [hunk 5]
```diff
@@ -399,9 +399,14 @@ String generateRandomRealmPath() {
 }
 
 final random = Random();
-String generateRandomString(int len) {
-  const chars = 'abcdefghjklmnopqrstuvwxuz';
-  return List.generate(len, (index) => chars[random.nextInt(chars.length)]).join();
+String generateRandomString(int len, {bool allowUnicode = false, String includingSymbols = ""}) {
```
Comment (blagoev) [inline test/test.dart:402]: This will work but I find the API little weird. 
I would prefer if the symbols passed are replacing the symbols used in the method. So the caller is either using the default implementation of `generateRandomString` or passing a character set specific at the call site. 
Otherwise one needs to really go and check which symbols are the default ones and also most importantly one can not exclude symbols from the default ones. Say for example I want to generate a random string which does not include `abc` in it.

Lets rename `includingSymbols` to characterSet and make it optional and null. If the caller passes it then we use only this characterSet to generate the random string otherwise we use the default one

**Review by blagoev — CHANGES_REQUESTED** [review]

File: test/test.dart (hunk 5 above)
Comment (blagoev) [reply test/test.dart:402]: or better yet we initialize the `characterSet` parameter with the default character set right in the method definition

Comment (desistefanova) [reply test/test.dart]: If we replace the whole charset with `characterSet `parameter's value then the parameter `allowUnicode` is useless. I will have to add again the method `generateRandomUnicodeString`. I refactored the methods according to your suggestions.

**Review by blagoev — APPROVED** [review]
